src/disco/gui/fd_gui_printf.c

```diff
@@ -0,0 +1,1172 @@
+#ifndef HEADER_fd_src_disco_gui_fd_gui_printf_h
+#define HEADER_fd_src_disco_gui_fd_gui_printf_h
+
+#include <ctype.h>
+#include <stdio.h>
+
+#include "fd_gui_printf.h"
+
+#include "../../ballet/http/fd_http_server_private.h"
+#include "../../ballet/utf8/fd_utf8.h"
+
+#define FD_GUI_PRINTF_DEBUG 0
+
+static void
+jsonp_strip_trailing_comma( fd_gui_t * gui ) {
+  if( FD_LIKELY( !gui->http->stage_err &&
+                 gui->http->stage_len>=1UL &&
+                 gui->http->oring[ (gui->http->stage_off%gui->http->oring_sz)+gui->http->stage_len-1UL ]==(uchar)',' ) ) {
+    gui->http->stage_len--;
+  }
+}
+
+static void
+jsonp_open_object( fd_gui_t *   gui,
+                   char const * key ) {
+  if( FD_LIKELY( key ) ) fd_http_server_printf( gui->http, "\"%s\":{", key );
+  else                   fd_http_server_printf( gui->http, "{" );
+}
+
+static void
+jsonp_close_object( fd_gui_t * gui ) {
+  jsonp_strip_trailing_comma( gui );
+  fd_http_server_printf( gui->http, "}," );
+}
+
+static void
+jsonp_open_array( fd_gui_t *   gui,
+                  char const * key ) {
+  if( FD_LIKELY( key ) ) fd_http_server_printf( gui->http, "\"%s\":[", key );
+  else                   fd_http_server_printf( gui->http, "[" );
+}
+
+static void
+jsonp_close_array( fd_gui_t * gui ) {
+  jsonp_strip_trailing_comma( gui );
+  fd_http_server_printf( gui->http, "]," );
+}
+
+static void
+jsonp_ulong( fd_gui_t *   gui,
+             char const * key,
+             ulong        value ) {
+  if( FD_LIKELY( key ) ) fd_http_server_printf( gui->http, "\"%s\":%lu,", key, value );
+  else                   fd_http_server_printf( gui->http, "%lu,", value );
+}
+
+static void
+jsonp_long( fd_gui_t *   gui,
+            char const * key,
+            long         value ) {
+  if( FD_LIKELY( key ) ) fd_http_server_printf( gui->http, "\"%s\":%lu,", key, value );
+  else                   fd_http_server_printf( gui->http, "%ld,", value );
+}
+
+static void 
+jsonp_double( fd_gui_t *   gui,
+              char const * key,
+              double       value ) {
+  if( FD_LIKELY( key ) ) fd_http_server_printf( gui->http, "\"%s\":%.2f,", key, value );
+  else                   fd_http_server_printf( gui->http, "%.2f,", value );
+}
+
+static void
+jsonp_sanitize_str( fd_http_server_t * http,
```

Comment:
I spent some time thinking about this code's behavior on unicode strings, and I think it's fine. All bytes that are part of a non-ASCII character have the high bit set, so this will never replace a byte in the middle of a multi-byte codepoint.